[FIX] hr_attendance_analysis various bugs

tafaRU opened the pull request:
This PR contains the changes made on https://code.launchpad.net/~agilebg/hr-timesheet/fix-hr_attendance_analysis-various-bugs before the project was hosted on github plus some others.
## Main changes
1. Merged branch lp:~agilebg/hr-timesheet/fix-hr_attendance_analysis-reference-is-not-defined and moved the related MP https://code.launchpad.net/~agilebg/hr-timesheet/fix-hr_attendance_analysis-reference-is-not-defined/+merge/221250
2. Merged branch lp:~agilebg/hr-timesheet/fix-hr_attendance_analysis-bracket and moved the related MP https://code.launchpad.net/~agilebg/hr-timesheet/fix-hr_attendance_analysis-bracket/+merge/221278
3. Merged branch lp:~agilebg/hr-timesheet/fix-hr_attendance_analysis-report-timezone and moved the related MP https://code.launchpad.net/~agilebg/hr-timesheet/fix-hr_attendance_analysis-report-timezone/+merge/221374
4. Merged branch lp:~agilebg/hr-timesheet/hr_attendance_analysis-imp-time_difference-message and moved the related MP https://code.launchpad.net/~agilebg/hr-timesheet/hr_attendance_analysis-imp-time_difference-message/+merge/225275
5. Deleted some unused methods and applied a bit of refactoring
6. Fixed the total leaves calculation: now the current_total_leaves value is calculated considering the working time and the timezone
7. Added hr_attendance_analysis/test/attendances.yml

Thank you.

eLBati committed:
[FIX]
File "/home/elbati/workspace/openerp/progetti/agilebg/git/agilebg/parts/hr-timesheet-7.0/hr_attendance_analysis/wizard/print_calendar_report.py", line 160, in print_calendar
    for calendar_attendance in reference.attendance_ids:
NameError: global name 'reference' is not defined

eLBati committed:
[FIX] wrong opening bracket due to PEP8 restyle:
http://bazaar.launchpad.net/~hr-core-editors/hr-timesheet/7.0/revision/40.2.21

[ADD] button_dummy to force recomputing attendace

eLBati committed:
[FIX] wrong opening bracket due to PEP8 restyle:
http://bazaar.launchpad.net/~hr-core-editors/hr-timesheet/7.0/revision/40.2.21

[ADD] button_dummy to force recomputing attendace

eLBati committed:
[ADD] 'Update all the attendance data' button on company

eLBati committed:
[ADD] 'Update all the attendance data' button on company

eLBati committed:
[FIX] hr_attendance_analysis
using correct timezone to print the sign in/out time

eLBati committed:
[FIX] hr_attendance_analysis
using correct timezone to print the sign in/out time

eLBati committed:
[FIX] hour and minutes

eLBati committed:
[FIX] hour and minutes

eLBati committed:
[FIX] problem with times that after been converted to correct timezone change its date.
For example: 29/05/2014 00:12 GMT+2 >> 28/05/2014 22:12 UTC

eLBati committed:
[FIX] problem with times that after been converted to correct timezone change its date.
For example: 29/05/2014 00:12 GMT+2 >> 28/05/2014 22:12 UTC

eLBati committed:
[FIX] comment

eLBati committed:
[FIX] comment

eLBati committed:
[IMP] time_difference error msg
This allows, when time diff errors are raised, to identify the raising attendance

eLBati committed:
[IMP] time_difference error msg
This allows, when time diff errors are raised, to identify the raising attendance

eLBati committed:
[FIX] syntax error

eLBati committed:
[FIX] syntax error

eLBati committed:
[imp] wizard error messages

eLBati committed:
[imp] wizard error messages

eLBati committed:
[IMP] wizard error message

eLBati committed:
[IMP] wizard error message

tafaRU committed:
[MRG] Merged upstream

tafaRU committed:
[IMP] Applied a bit of refactoring, added methods about centered_attendance calculation

tafaRU committed:
[IMP] Applied a bit of refactoring, added methods about centered_attendance calculation

tafaRU committed:
[MRG] Merged branch lp:~agilebg/hr-timesheet/fix-hr_attendance_analysis-reference-is-not-defined

Fixed "NameError: global name 'reference' is not defined"

tafaRU committed:
[MRG] Merged branch lp:~agilebg/hr-timesheet/fix-hr_attendance_analysis-bracket

Fixed wrong opening bracket due to PEP8 restyle:
    http://bazaar.launchpad.net/~hr-core-editors/hr-timesheet/7.0/revision/40.2.21

Added button_dummy to force recomputing attendance (on attendance form and company form)

tafaRU committed:
[MRG] Merged branch lp:~agilebg/hr-timesheet/fix-hr_attendance_analysis-report-timezone

Fixed hr_attendance_analysis using correct timezone to print the sign in/out time

tafaRU committed:
[MRG] Merged branch lp:~agilebg/hr-timesheet/hr_attendance_analysis-imp-time_difference-message

Improved time_difference message: this change allows, when error is raised, to identify the raising attendance

tafaRU committed:
[IMP] Applied a bit of refactoring, added methods about centered_attendance calculation

tafaRU committed:
[IMP] Added get_matched_schedule method

tafaRU committed:
[IMP] Added get_matched_schedule method

tafaRU committed:
[REM] pdb

tafaRU committed:
[REM] pdb

tafaRU committed:
[REM] unused methods of hr.attendance

tafaRU committed:
[REM] unused methods of hr.attendance

tafaRU committed:
[IMP] method _split_no_recursive_attendance

tafaRU committed:
[IMP] method _split_no_recursive_attendance

tafaRU committed:
[IMP] Applied some of refactoring.

Renamed the following hr.attendance methods:

_split_no_recursive_attendance --> split_interval_time_by_precision
attendance_to_hour --> datetime_to_hour
centered_attendance --> mid_time_interval

tafaRU committed:
[IMP] Applied some of refactoring.

Renamed the following hr.attendance methods:

_split_no_recursive_attendance --> split_interval_time_by_precision
attendance_to_hour --> datetime_to_hour
centered_attendance --> mid_time_interval

tafaRU committed:
[FIX] Total leaves calculation: now the current_total_leaves value is calculated considering the working time and the timezone

tafaRU committed:
[FIX] Total leaves calculation: now the current_total_leaves value is calculated considering the working time and the timezone

tafaRU committed:
[IMP] Indentation of if condition

eLBati commented:
Thanks @tafaRU , could you also point out the reviews already done on the old MPs?

tafaRU commented:
@eLBati I've just updated the description including the related MPs moved.

eLBati commented on hr_attendance_analysis/wizard/print_calendar_report.py:
```diff
@@ -200,17 +233,58 @@ def print_calendar(self, cr, uid, ids, context=None):
+                    date_from = date_from.replace(
+                        tzinfo=pytz.utc).astimezone(active_tz)
+                    date_to = date_to.replace(
+                        tzinfo=pytz.utc).astimezone(active_tz)
```
Why conversion to user timezone?
Is `matched_schedule` aware of timezone?

tafaRU replied:
date_from and date_to are not aware of timezone so I thought to convert them as we made for the attendances.

eLBati replied:
After investiganting, I saw the working schedule is not converted to timezone. That is, if your working schedule is from 09:00 to 13:00 (as you see it in the web interface), in the DB the values are exactly 9 and 13.
On the other hand, the leaves (as well as the attendances) are always saved as UTC in the DB. That is, if you leave at 09:00 and your timezone is GTM+2, the DB will contain 07:00.
So, the conversion here is correct
:+1:

eLBati committed:
[ADD] first minimal test

eLBati committed:
[ADD] first minimal test

eLBati committed:
[ADD] print calendar test

eLBati committed:
[ADD] print calendar test

eLBati committed:
[ADD] several tests

eLBati committed:
[ADD] several tests

eLBati committed:
[FIX] wrong opening bracket due to PEP8 restyle:
    http://bazaar.launchpad.net/~hr-core-editors/hr-timesheet/7.0/revision/40.2.21

eLBati committed:
[FIX] wrong opening bracket due to PEP8 restyle:
    http://bazaar.launchpad.net/~hr-core-editors/hr-timesheet/7.0/revision/40.2.21

eLBati commented on hr_attendance_analysis/hr_attendance.py:
```diff
@@ -293,11 +316,13 @@ def _get_attendance_duration(self, cr, uid, ids, field_name, arg, context=None):
-                                    )
+                                    self.datetime_to_hour(attendance_start)
+                                )
                                 if attendance_start_hour >= (
```
Why the fix from https://code.launchpad.net/~agilebg/hr-timesheet/fix-hr_attendance_analysis-bracket/+merge/221278 is missing here and at line 343?

eLBati committed:
[ADD] report calendar test for 2nd day

eLBati committed:
[ADD] report calendar test for 2nd day

eLBati committed:
[IMP] test with timezone

eLBati committed:
[IMP] test with timezone

eLBati committed:
[FIX] remove timezone test because does not make sense without interface

eLBati committed:
[FIX] remove timezone test because does not make sense without interface

tafaRU committed:
Merge pull request #1 from eLBati/imp_fix-hr_attendance_analysis-various-bugs

[ADD] tests and fix for 'wrong opening bracket'

bwrsandman committed:
Merge pull request #6 from pedrobaeza/7.0-fix_pep8

[FIX] PEP8

Fixes #3

pedrobaeza commented:
Please rebase to have last PEP8 changes included. You might have to resolve conflicts.

eLBati committed:
Merge branch 'imp_fix-hr_attendance_analysis-various-bugs' of github.com:eLBati/hr-timesheet into imp_fix-hr_attendance_analysis-various-bugs

Conflicts:
	hr_attendance_analysis/hr_attendance.py
	hr_attendance_analysis/wizard/print_calendar_report.py

eLBati commented:
@tafaRU rebased at https://github.com/tafaRU/hr-timesheet/pull/2

tafaRU committed:
Merge pull request #2 from eLBati/imp_fix-hr_attendance_analysis-various-bugs

[REBASE] Imp fix hr attendance analysis various bugs

pedrobaeza commented:
There are some PEP8 errors introduced in this PR.

tafaRU committed:
[FIX] hr_attendance.py: removed undefine name 'attr', 'attr_start'

tafaRU committed:
[FIX] hr_attendance.py: PEP8 cleanup

tafaRU committed:
[FIX] wizard/print_calendar_report.py: PEP8 cleanup

tafaRU committed:
[FIX] Applied PEP8 cleanup running flake8

tafaRU committed:
[FIX] test/attendances.yml
I fixed the attendance names to avoid conflicts with those attendances already existing for
the employee 'hr.employee_fp', loaded in demo data using by the file hr_attendance/hr_attendance_demo.xml

tafaRU commented:
@pedrobaeza I fixed the PEP8 errors and the test/attendances.yml.
Now the Travis CI light is green colored :grin:

rvalyi commented:
:+1:  LGTM

bwrsandman commented:
:+1:

pedrobaeza commented:
Thank you very much.

:+1: 

Regards.
